Documentation/ExtensionArchitecture/Extbase/Reference/FileUpload.rst

```diff
@@ -0,0 +1,489 @@
+..  include:: /Includes.rst.txt
+
+..  index:: Extbase, FileUpload
+
+..  _extbase_fileupload:
+
+===========
+File upload
+===========
+
+Implementing file uploads / attachments to Extbase Domain Models
+has always been a bit of a challenge.
+
+While it is easy to access an existing file reference in a Domain Model,
+writing new files to the :ref:`FAL (File Access Layer) <t3coreapi:using-fal>`
+takes more effort.
+
+..  _extbase_fileupload_accessing:
+
+Accessing a file reference in an Extbase Domain Model
+-----------------------------------------------------
+
+You need two components for the structural information: The Domain
+Model definition and the TCA entry.
+
+..  literalinclude:: _FileUpload/_Blog.php
+    :caption: EXT:my_extension/Classes/Domain/Model/Blog.php
+
+..  literalinclude:: _FileUpload/_tx_myextension_domain_model_blog.php
+    :caption: EXT:my_extension/Configuration/TCA/tx_myextension_domain_model_blog.php
+
+Once this is set up, you can create/edit records through the TYPO3
+backend, attach a single or multiple files in it. Then using a normal
+controller and fluid template, you can display an image:
+
+..  literalinclude:: _FileUpload/_BlogController.php
+    :caption: EXT:my_extension/Classes/Controller/BlogController.php
+
+..  literalinclude:: _FileUpload/_Show.html
+    :caption: EXT:my_extension/Resources/Private/Templates/Blog/Show.html
+
+On the PHP-side within controllers, you can use the usual
+:php:`$blogItem->getSingleFile()` and :php:`$blogItem->getMultipleFilers()`
+Extbase getters to retrieve the FileReference object.
+
+..  _extbase_fileupload_writing:
+
+Writing FileReference entries
+-----------------------------
+
+..  _extbase_fileupload_writing-manual:
+
+Manual handling
+...............
+
+With TYPO3 versions below v13, attaching files to an Extbase Domain Model
+was only possible by either:
+
+*  Manually evaluate the :php:`$_FILES` data, process and validate the data,
+   use raw QueryBuilder write actions on :sql:`sys_file` and :sql:`sys_file_reference`
+   to persist the files quickly, or use at least some API methods:
+
+    ..  literalinclude:: _FileUpload/_ApiUpload.php
+        :caption: EXT:my_extension/Classes/Controller/BlogController.php, excerpt
+
+   Instead of raw access to :php:`$_FILES`, starting with TYPO3 v12 the recommendation
```

Comment:
Above you write before v13, now starting with v12? So, what was really used in v12?

Reply:
UploadedFile exists since v12. Extbase File Uploads since 13.3 - those are two distinct features. It's a bit complex sadly.

UploadedFile is mainly used by EXT:form frontend, but not in Extbase domain models.